app/upgrades/v20/upgrades.go

```diff
@@ -134,133 +136,512 @@ func InitializeLastProviderConsensusValidatorSet(
+	// in legacy proposals (which have only one message)
+	messages := proposal.GetMessages()
+	if len(messages) != 1 {
+		return nil
+	}
+	msg := messages[0]
+
+	// ignore non-legacy proposals
+	sdkLegacyMsg, isLegacyProposal := msg.GetCachedValue().(*govtypes.MsgExecLegacyContent)
+	if !isLegacyProposal {
+		return nil
+	}
+	content, err := govtypes.LegacyContentFromMessage(sdkLegacyMsg)
+	if err != nil {
+		return err
+	}
 
-		msgRemove, ok := content.(*providertypes.ConsumerRemovalProposal)
-		if ok {
-			anyMsg, err := MigrateLegacyConsumerRemoval(*msgRemove, govKeeper.GetAuthority())
+	switch msg := content.(type) {
+	case *providertypes.ConsumerAdditionProposal:
+		return MigrateConsumerAdditionProposal(
+			ctx,
+			msgServer,
+			providerKeeper,
+			govKeeper,
+			proposal,
+			msg,
+		)
+
+	case *providertypes.ConsumerRemovalProposal:
+		return MigrateConsumerRemovalProposal(
+			ctx,
+			msgServer,
+			providerKeeper,
+			govKeeper,
+			proposal,
+			msg,
+		)
+
+	case *providertypes.ConsumerModificationProposal:
+		return MigrateConsumerModificationProposal(
+			ctx,
+			msgServer,
+			providerKeeper,
+			govKeeper,
+			proposal,
+			msg,
+		)
+
+	case *providertypes.ChangeRewardDenomsProposal:
+		return MigrateChangeRewardDenomsProposal(
+			ctx,
+			msgServer,
+			providerKeeper,
+			govKeeper,
+			proposal,
+			msg,
+		)
+	}
+
+	return nil
+}
+
+// MigrateConsumerAdditionProposal migrates a ConsumerAdditionProposal
+func MigrateConsumerAdditionProposal(
+	ctx sdk.Context,
+	msgServer providertypes.MsgServer,
+	providerKeeper providerkeeper.Keeper,
+	govKeeper govkeeper.Keeper,
+	proposal govtypes.Proposal,
+	msg *providertypes.ConsumerAdditionProposal,
+) error {
+	if proposal.Status == govtypes.StatusPassed {
+		// ConsumerAdditionProposal that passed -- it was added to the
+		// list of pending consumer addition proposals, which was deleted during
+		// the migration of the provider module
+		for _, consumerID := range providerKeeper.GetAllActiveConsumerIds(ctx) {
+			chainID, err := providerKeeper.GetConsumerChainId(ctx, consumerID)
 			if err != nil {
-				return err
+				return err // this means something is wrong with the provider state
 			}
-			proposal.Messages[idx] = anyMsg
-			continue // skip the rest of the loop
+			if chainID == msg.ChainId {
+				// this proposal was already handled in a previous block
+				ctx.Logger().Info(
+					fmt.Sprintf(
+						"Proposal with ID(%d) was skipped as it was already handled - consumerID(%s), chainID(%s), spawnTime(%s)",
+						proposal.Id, consumerID, msg.ChainId, msg.SpawnTime.String(),
+					),
+				)
+				return nil
+			}
+		}
+		// create a new consumer chain with all the parameters
+		metadata, err := metadataFromCAP(msg)
+		if err != nil {
+			return err
+		}
+		initParams, err := initParamsFromCAP(msg)
+		if err != nil {
+			return err
 		}
+		powerShapingParams, err := powerShapingParamsFromCAP(msg)
+		if err != nil {
```

Comment:
```suggestion
		powerShapingParams, err := powerShapingParamsFromCAP(msg)
```

Reply:
@sainoe I don't understand you suggestion.